### PR TITLE
Drop support for Python 3.7

### LINKED ISSUE #76: Seems like Numpy has dropped python 3.7 in v1.23.5
### Description

https://github.com/numpy/numpy/blob/v1.23.5/setup.py#L453

### PR DESCRIPTION
**Pull request by ricardoV94**
In addition:
* Make annotations compatible with oldest supported Python version (mypy was failing on Python 3.8)
* Test pre-commit in all supported versions
* Updated isort dependency

Closes #76

### CODE REVIEW

**Review by twiecki — APPROVED** [review]

**Comment by ricardoV94** [comment]
Codecov upload failed... dunno if it's a fluke

File: pytensor/link/numba/dispatch/elemwise_codegen.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+from __future__ import annotations
```
Comment (ferrine) [inline pytensor/link/numba/dispatch/elemwise_codegen.py:1]: Why is this needed?

**Comment by Armavica** [comment]
Should we maybe add `pyupgrade` to `pre-commit`, like for PyMC?

File: pytensor/link/numba/dispatch/elemwise_codegen.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/link/numba/dispatch/elemwise_codegen.py:1]: Otherwise in Python 3.8 you get `pytensor/link/numba/dispatch/elemwise_codegen.py:72: error: "list" is not subscriptable, use "typing.List" instead`

Comment (ricardoV94) [reply pytensor/link/numba/dispatch/elemwise_codegen.py:1]: I'm okay either way, but that's why I changed it to run pre-commit in all supported versions. We missed this issue because it was running only in 3.9

**Comment by ricardoV94** [comment]
> Should we maybe add `pyupgrade` to `pre-commit`, like for PyMC?

Yup. Mind opening an issue / PR, or do you think it should be done here?

**Comment by Armavica** [comment]
Yeah no worries, I can create a separate PR.

**Review by michaelosthege — APPROVED** [review]

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -50,11 +50,14 @@ jobs:
     needs: changes
     runs-on: ubuntu-latest
     if: ${{ needs.changes.outputs.changes == 'true' }}
+    strategy:
+      matrix:
+        python-version: ["3.8", "3.9", "3.10"]
```
[the remark this answers is not in the document]
Comment (michaelosthege) [reply .github/workflows/test.yml:55]: AFAIK numba is not yet available on 3.11 but that should not stop us.
Looking forward to all these speed ups

[the remark this answers is not in the document]
Comment (ricardoV94) [reply .github/workflows/test.yml:55]: There's another PR for 3.11. We could test support for ot t but would need to tweak jobs so that they run all tests but the numba ones.

I would rather do it in that PR. https://github.com/pymc-devs/pytensor/pull/198

Pushes there are welcome
